plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/ptf/RemoteQuery.java

```diff
@@ -0,0 +1,120 @@
+import io.trino.plugin.jdbc.JdbcColumnHandle;
+import io.trino.plugin.jdbc.JdbcMetadata;
+import io.trino.plugin.jdbc.JdbcTableHandle;
+import io.trino.plugin.jdbc.JdbcTransactionManager;
+import io.trino.plugin.jdbc.PreparedQuery;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.connector.ConnectorTableHandle;
+import io.trino.spi.connector.ConnectorTransactionHandle;
+import io.trino.spi.ptf.Argument;
+import io.trino.spi.ptf.ConnectorTableFunction;
+import io.trino.spi.ptf.ConnectorTableFunctionHandle;
+import io.trino.spi.ptf.Descriptor;
+import io.trino.spi.ptf.Descriptor.Field;
+import io.trino.spi.ptf.ScalarArgument;
+import io.trino.spi.ptf.ScalarArgumentSpecification;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.trino.spi.ptf.DescriptorMapping.EMPTY_MAPPING;
+import static io.trino.spi.ptf.ReturnTypeSpecification.GenericTable.GENERIC_TABLE;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public class RemoteQuery
+        implements Provider<ConnectorTableFunction>
+{
+    public static final String NAME = "remote_query";
+
+    private final JdbcTransactionManager transactionManager;
+
+    @Inject
+    public RemoteQuery(JdbcTransactionManager transactionManager)
+    {
+        this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+    }
+
+    @Override
+    public ConnectorTableFunction get()
+    {
+        // TODO wrap in ClassLoaderSafeConnectorTableFunction? (see also TestClassLoaderSafeWrappers)
+        return new RemoteQueryFunction(transactionManager);
+    }
+
+    public static class RemoteQueryFunction
+            extends ConnectorTableFunction
+    {
+        public static final String SCHEMA_NAME = "system";
+
+        private final JdbcTransactionManager transactionManager;
+
+        public RemoteQueryFunction(JdbcTransactionManager transactionManager)
+        {
+            super(SCHEMA_NAME, NAME, List.of(new ScalarArgumentSpecification("query", VARCHAR)), GENERIC_TABLE);
+            this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+        }
+
+        @Override
+        public Analysis analyze(ConnectorSession session, ConnectorTransactionHandle transaction, Map<String, Argument> arguments)
+        {
+            ScalarArgument argument = (ScalarArgument) getOnlyElement(arguments.values());
+            String query = ((Slice) argument.getValue()).toStringUtf8();
+            PreparedQuery preparedQuery = new PreparedQuery(query, ImmutableList.of());
+
+            JdbcMetadata metadata = transactionManager.getMetadata(transaction);
+            JdbcTableHandle tableHandle = metadata.getTableHandle(session, preparedQuery);
```

Comment:
Do we actually need to use JdbcTableHandle here? Maybe we could introduce separate table handle class?

Reply:
But it'd would require more changes on the JdbcRecordSetProvider to handle it

Reply:
as discussed offline. `JdbcTableHandle` is really a good fit here, because it already represents "arbitrary relation"